{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/chaeeun/.conda/envs/biomedlm/lib/python3.9/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "cuda\n"
     ]
    }
   ],
   "source": [
    "import os \n",
    "# os.environ[\"CUDA_DEVICE_ORDER\"]=\"PCI_BUS_ID\"\n",
    "# os.environ[\"CUDA_VISIBLE_DEVICES\"]=\"1, 2\"\n",
    "\n",
    "import torch\n",
    "print(torch.backends.cuda.is_built())\n",
    "print(torch.cuda.is_available())\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "print(device)\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GPT2LMHeadModel(\n",
       "  (transformer): GPT2Model(\n",
       "    (wte): Embedding(28896, 2560)\n",
       "    (wpe): Embedding(1024, 2560)\n",
       "    (drop): Dropout(p=0.1, inplace=False)\n",
       "    (h): ModuleList(\n",
       "      (0): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (1): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (2): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (3): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (4): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (5): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (6): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (7): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (8): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (9): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (10): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (11): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (12): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (13): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (14): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (15): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (16): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (17): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (18): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (19): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (20): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (21): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (22): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (23): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (24): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (25): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (26): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (27): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (28): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (29): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (30): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "      (31): GPT2Block(\n",
       "        (ln_1): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (attn): GPT2Attention(\n",
       "          (c_attn): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (attn_dropout): Dropout(p=0.1, inplace=False)\n",
       "          (resid_dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "        (ln_2): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "        (mlp): GPT2MLP(\n",
       "          (c_fc): Conv1D()\n",
       "          (c_proj): Conv1D()\n",
       "          (act): NewGELUActivation()\n",
       "          (dropout): Dropout(p=0.1, inplace=False)\n",
       "        )\n",
       "      )\n",
       "    )\n",
       "    (ln_f): LayerNorm((2560,), eps=1e-05, elementwise_affine=True)\n",
       "  )\n",
       "  (lm_head): Linear(in_features=2560, out_features=28896, bias=False)\n",
       ")"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from transformers import AutoConfig, AutoModelForSeq2SeqLM, AutoModelForCausalLM, AutoTokenizer, GPT2LMHeadModel, GPT2Tokenizer\n",
    "import json\n",
    "\n",
    "# model_name = 'google/flan-t5-small'\n",
    "# model_name = 'google/flan-t5-base'\n",
    "# model_name = 'google/flan-t5-small'\n",
    "# model_name = 'google/flan-t5-xl'\n",
    "# model_name = 'google/flan-t5-large'\n",
    "# model_name = 'google/flan-t5-xxl'\n",
    "# model_name = \"google/pegasus-pubmed\"\n",
    "# model_name = 'microsoft/BioGPT-Large'\n",
    "\n",
    "model_name = \"stanford-crfm/BioMedLM\"\n",
    "\n",
    "\n",
    "# config.n_positions = 2048\n",
    "\n",
    "\n",
    "if model_name == \"microsoft/BioGPT-Large\":\n",
    "    config = AutoConfig.from_pretrained(model_name)\n",
    "    model = AutoModelForCausalLM.from_pretrained(model_name, config=config)\n",
    "elif model_name == \"stanford-crfm/BioMedLM\":\n",
    "    config = AutoConfig.from_pretrained(model_name)\n",
    "    model = GPT2LMHeadModel.from_pretrained(model_name, config=config)\n",
    "    tokenizer = GPT2Tokenizer.from_pretrained(model_name)\n",
    "\n",
    "else:\n",
    "    config = AutoConfig.from_pretrained(model_name)\n",
    "    model = AutoModelForSeq2SeqLM.from_pretrained(model_name, config=config)\n",
    "    tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "    \n",
    "model.to(device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPT2Config {\n",
      "  \"_name_or_path\": \"stanford-crfm/BioMedLM\",\n",
      "  \"activation_function\": \"gelu_new\",\n",
      "  \"architectures\": [\n",
      "    \"GPT2LMHeadModel\"\n",
      "  ],\n",
      "  \"attn_pdrop\": 0.1,\n",
      "  \"bos_token_id\": 28895,\n",
      "  \"embd_pdrop\": 0.1,\n",
      "  \"eos_token_id\": 28895,\n",
      "  \"initializer_range\": 0.02,\n",
      "  \"layer_norm_epsilon\": 1e-05,\n",
      "  \"model_type\": \"gpt2\",\n",
      "  \"n_ctx\": 1024,\n",
      "  \"n_embd\": 2560,\n",
      "  \"n_head\": 20,\n",
      "  \"n_inner\": null,\n",
      "  \"n_layer\": 32,\n",
      "  \"n_positions\": 1024,\n",
      "  \"reorder_and_upcast_attn\": false,\n",
      "  \"resid_pdrop\": 0.1,\n",
      "  \"scale_attn_by_inverse_layer_idx\": true,\n",
      "  \"scale_attn_weights\": true,\n",
      "  \"summary_activation\": null,\n",
      "  \"summary_first_dropout\": 0.1,\n",
      "  \"summary_proj_to_labels\": true,\n",
      "  \"summary_type\": \"cls_index\",\n",
      "  \"summary_use_proj\": true,\n",
      "  \"task_specific_params\": {\n",
      "    \"text-generation\": {\n",
      "      \"do_sample\": true,\n",
      "      \"max_length\": 50\n",
      "    }\n",
      "  },\n",
      "  \"torch_dtype\": \"float32\",\n",
      "  \"transformers_version\": \"4.24.0\",\n",
      "  \"use_cache\": false,\n",
      "  \"vocab_size\": 28896\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## experimenting for BioMedLM. Delete this cell later. \n",
    "\n",
    "config = AutoConfig.from_pretrained(model_name)\n",
    "print(config)\n",
    "\n",
    "tokenizer.eos_token_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Models to run\n",
    "# bioGPT\n",
    "# bioGPT large\n",
    "# Flan-T5\n",
    "# google/pegasus-pubmed\n",
    "\n",
    "# ['base', 'FG_template', 'evidence-only']\n",
    "\n",
    "num_shots = 0\n",
    "mode = 'base'\n",
    "# instruction = None\n",
    "instruction = \"Answer the query: \"\n",
    "# instruction = \"Given the evidence, answer the query provided: \"\n",
    "# instruction = \"Based on the evidence provided, what is your conclusion regarding the query? \"\n",
    "# instruction = \"Write a summary of the evidence and query in your own words: \"\n",
    "# instruction = \"What are the potential implications of the given evidences in relation to the query? \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# outputs propt template with tags to be replaced\n",
    "# could avoid the problem of overlapping with query by splitting contexts before passing it to the two functions \n",
    "import random\n",
    "# (contexts, queries, mode, num_shots, instruction=None):\n",
    "class Prompt:\n",
    "    def __init__(self, contexts, queries):\n",
    "        self.contexts = contexts\n",
    "        self.queries = queries\n",
    "        # self.mode = mode \n",
    "        # self.num_shots = num_shots\n",
    "            \n",
    "    def get_context_template(self, contexts_subset, num_examples): # or you could return all possible templates as a list \n",
    "        \n",
    "        if num_examples==0:\n",
    "            return None\n",
    "            \n",
    "        # prefix on every example, suffix only at the very end of context template\n",
    "        example_part = \"\"\n",
    "        for idx_example in range(num_examples): # change this part for random sampling / all permutation / etc\n",
    "            example = contexts_subset[idx_example]\n",
    "            example_part += \"<EX_PREFIX>\"\n",
    "            for i, ev in enumerate(example['evidences']):\n",
    "                example_part += f\"evidence {i}: {ev} \\n\"\n",
    "            \n",
    "            example_part += f\"query: {example['query']} \\n\"\n",
    "            example_part += \"<EX_OUTPUT>\" # either Output: or Instruction or both?\n",
    "            example_part += f\"{example['ground_truth']} \\n\"\n",
    "        example_part += \"<EX_SUFFIX>\"\n",
    "            \n",
    "        return example_part\n",
    "\n",
    "    def get_query_templates(self, queries_subset):\t# returns a list of all possible query templates\n",
    "        \n",
    "        templates = []\n",
    "        # dict_3 = {**dict_1,**dict_2}\n",
    "        \n",
    "        for query_idx in range(len(queries_subset)):\n",
    "            query = queries_subset[query_idx]\n",
    "\n",
    "            ## Comment out the below three lines for no-evidence generation\n",
    "            # query_part = \"<QR_PREFIX>\\n\" # probs will just get rid of it \n",
    "            # for i, ev in enumerate(query['evidences']):\n",
    "            #     query_part += f\"evidence {i}: {ev} \\n\"\n",
    "\n",
    "            query_part = \"\"\n",
    "            query_part += f\"query: {query['query']} \\n\"\n",
    "            query_part += \"<QR_SUFFIX>\" # probs instruction\n",
    "            \n",
    "            templates.append(query_part)\n",
    "            \n",
    "        return templates\n",
    "\n",
    "    def transform_templates(self, context_template, query_templates, mode, instruction):\n",
    "\n",
    "        ## FG_template이면 query가 없고 evidence-only면 ... 뭐가 없지? 얘는 context를 날리는게 맞는 듯. \n",
    "        \n",
    "        assert (mode in ['base', 'FG_template', 'evidence-only']), \"invalid mode\"\n",
    "        \n",
    "        instruction = (\"Summarize given evidences and query: \" if instruction==None else instruction)\n",
    "        preface = \"I gave a friend an instruction and five inputs. The friend read the instruction and wrote an output for every one of the inputs. Here are the input-output pairs: \\n\"\n",
    "        \n",
    "        if mode==\"base\":\n",
    "            ex_prefix, ex_output, ex_suffix, qr_prefix, qr_suffix = \"\", instruction, \"\", \"\", instruction\n",
    "        elif mode==\"FG_template\": # query part irrelevant\n",
    "            ex_prefix, ex_output, ex_suffix = (preface + \"\\n\" + \"Input: \\n\"), \"Output: \\n\", \"\\nThe instruction was: \"\n",
    "        elif mode==\"evidence-only\": # context part irrelevant\n",
    "            context_template = None\n",
    "            qr_prefix, qr_suffix = \"\", \"\",\n",
    "            \n",
    "        # \n",
    "        context_transformed = context_template.replace('<EX_PREFIX>', ex_prefix).replace('<EX_OUTPUT>', ex_output).replace('<EX_SUFFIX>', ex_suffix) if (context_template!=None) else \"\"\n",
    "        \n",
    "        if mode=='FG_template':\n",
    "            queries_transformed = None\n",
    "        else:\n",
    "            queries_transformed = []\n",
    "            for query_template in query_templates:\n",
    "                temp = query_template.replace('<QR_PREFIX>', qr_prefix).replace('<QR_SUFFIX>', qr_suffix)\n",
    "                queries_transformed.append(temp)\n",
    "            \n",
    "        return context_transformed, queries_transformed\n",
    "        \n",
    "\n",
    "\n",
    "    def get_prompt(self, mode, num_shots, instruction=None):\n",
    "        \n",
    "\n",
    "        # randomly select from contexts num_shots contexts\n",
    "        indices = list(range(len(self.contexts)))  # create a list of indices from 0 to 9\n",
    "        context_indices = random.sample(indices, num_shots)\n",
    "        contexts = [self.contexts[i] for i in context_indices]\n",
    "        as_queries = [self.contexts[i] for i in range(len(self.contexts)) if i not in context_indices]\n",
    "        queries = self.queries + as_queries\n",
    "        \n",
    "        context_template = self.get_context_template(contexts, num_shots)\n",
    "        query_templates = self.get_query_templates(queries)\n",
    "        \n",
    "        context_transformed, queries_transformed = self.transform_templates(context_template, query_templates, mode, instruction)\n",
    "        \n",
    "        prompts = []\n",
    "        if mode != 'FG_template':\n",
    "            for query_transformed in queries_transformed:\n",
    "                prompt = context_transformed + query_transformed\n",
    "                prompts.append(prompt)\n",
    "        else:\n",
    "            prompts.append(context_transformed) # if mode is FG_template\n",
    "        \n",
    "        return prompts\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## for Google Sheets \n",
    "\n",
    "item1 = '''Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients. All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms.'''\n",
    "item2 = '''Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18'''\n",
    "item3 = '''CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of vSMC . maturation of vSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of vSMC.'''\n",
    "\n",
    "item3 = \"There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \"\n",
    "item4 = \"We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \"\n",
    "item5 = \"Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large\"\n",
    "item6 = \"Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK.\"\n",
    "item7 = \"Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-κB. A positive regulation between NOTCH and NF-κB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \"\n",
    "\n",
    "gt = \"Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGF-b1 signaling.\"\n",
    "\n",
    "set1 = {\"query\": 'What is the cause of the CADASIL?', \"evidences\": [item1, item2, item3]}\n",
    "set2 = {\"query\": 'What is the therapeutic target for CADASIL?', \"evidences\": [item3, item4, item5, item6, item7], \"ground_truth\": gt}\n",
    "\n",
    "# test_args = [(set2, set1),]\n",
    "\n",
    "contexts = [set2, ]\n",
    "queries = [set1, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('examples_20230207.json') as f:\n",
    "    loaded_dic = json.load(f)\n",
    "    contexts = loaded_dic['contexts']\n",
    "    queries = loaded_dic['queries']\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt_setup = Prompt(contexts, queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['query: What is the cause of the CADASIL? \\nGiven the evidence, answer the query provided: ', 'query: What is the therapeutic target for CADASIL? \\nGiven the evidence, answer the query provided: ']\n"
     ]
    }
   ],
   "source": [
    "# assert (mode in ['base', 'FG_template', 'evidence-only']), \"invalid mode\"\n",
    "\n",
    "# num_shots = 3\n",
    "# mode = 'FG_template'\n",
    "# instruction = ''\n",
    "\n",
    "prompts = prompt_setup.get_prompt(mode=mode, num_shots=num_shots, instruction=instruction)\n",
    "\n",
    "print(prompts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gpt2': 1024,\n",
       " 'gpt2-medium': 1024,\n",
       " 'gpt2-large': 1024,\n",
       " 'gpt2-xl': 1024,\n",
       " 'distilgpt2': 1024}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tokenizer.max_model_input_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## in progress. simplifying generation configuration process\n",
    "\n",
    "generate_kwargs = {\n",
    "    \"max_length\": 512,\n",
    "    # \"min_length\": 1024,\n",
    "    'do_sample': True, \n",
    "    'top_k': 50,\n",
    "    'num_beams': 5,   \n",
    "    'early_stopping': True,\n",
    "}\n",
    "\n",
    "tokenizer_kwargs = {\n",
    "    \"max_length\": 1024,\n",
    "    \"return_tensors\": \"pt\", # biomed by default only specifies this\n",
    "    \"truncation\": True,  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:28895 for open-end generation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cwd = /ssd1/chaeeun/biochatgpt_generation\n",
      "< PROMPT >\n",
      "\n",
      "query: What is the cause of the CADASIL? \n",
      "Given the evidence, answer the query provided: \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:28895 for open-end generation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['query: What is the cause of the CADASIL? \\nGiven the evidence, answer the query provided: \\n  -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Given the evidence, answer the query provided: \n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['query: What is the therapeutic target for CADASIL? \\nGiven the evidence, answer the query provided: \\nIntroduction {#sec1-1}\\n============\\n\\nDiabetes mellitus (DM) is one of the most common chronic diseases in the world.\\\\[[@ref1][@ref2][@ref3][@ref4][@ref5][@ref6][@ref7][@ref8][@ref9][@ref10][@ref11][@ref12][@ref13][@ref14][@ref15][@ref16][@ref17][@ref18][@ref19][@ref20][@ref21][@ref22][@ref23][@ref24][@ref25][@ref26][@ref27][@ref28][@ref29][@ref30][@ref31][@ref32][@ref33][@ref34][@ref35][@ref36][@ref37][@ref38][@ref39][@ref40][@ref41][@ref42][@ref43][@ref44][@ref45][@ref46][']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n"
     ]
    }
   ],
   "source": [
    "num_to_word = {}\n",
    "num_to_word[0]='zero-shot'\n",
    "num_to_word[1]='one-shot'\n",
    "num_to_word[2]='two-shot'\n",
    "num_to_word[3]='few-shot'\n",
    "\n",
    "output_file_name = f\"{model_name.replace('/', '-')}_{mode}-mode_{num_to_word[num_shots]}\"\n",
    "output_dic = {\n",
    "    'output' : [],\n",
    "    'config' : {'mode': mode, 'num_shots': num_shots}\n",
    "}\n",
    "\n",
    "\n",
    "# cwd = os.path.dirname(os.path.abspath(__file__))\n",
    "cwd = os.getcwd()\n",
    "print(f'cwd = {cwd}')\n",
    "file_name_json = os.path.join(cwd, f\"{output_file_name}.json\")\n",
    "\n",
    "with open(f'{output_file_name}.txt', \"w\") as results: # mujeen_\n",
    "    results.write(f'<< Model: {model_name} >>\\n\\n')\n",
    "\n",
    "    for prompt in prompts:\n",
    "\n",
    "        print('< PROMPT >\\n')\n",
    "        print(prompt)\n",
    "        results.write('< PROMPT >\\n\\n')\n",
    "        results.write(prompt)\n",
    "        \n",
    "        # inputs = tokenizer(prompt, return_tensors=\"pt\", max_length=1024, truncation=True)\n",
    "        inputs = tokenizer(prompt, **tokenizer_kwargs)\n",
    "        inputs.to(device)\n",
    "        # outputs = model.generate(**inputs, max_length=2048, min_length=100, num_beams=5)\n",
    "        outputs = model.generate(**inputs, **generate_kwargs)\n",
    "        generated = tokenizer.batch_decode(outputs, skip_special_tokens=True)\n",
    "        # print(generated)\n",
    "        # generated = *generated\n",
    "        \n",
    "        output_dic['output'].append({'prompt': prompt, 'generated': generated})\n",
    "\n",
    "        print('\\n\\n< GENERATED >\\n')\n",
    "        print(generated)\n",
    "        print('\\n###############################################################################################################################################################\\n')\n",
    "        results.write('\\n\\n< GENERATED >\\n\\n')\n",
    "        results.write(*generated)\n",
    "        results.write('\\n\\n###############################################################################################################################################################\\n')\n",
    "    \n",
    "\n",
    "    with open(file_name_json, \"w\") as outfile: \n",
    "        json.dump(output_dic, outfile)\n",
    "    outfile.close()\n",
    "results.close()\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling. \n",
      "\n",
      "evidence 0: CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. \n",
      "evidence 1: Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing. \n",
      "evidence 2: Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL. \n",
      "query: can nilvadipine help alleviate cadasil symptoms? \n",
      "Summarize given evidences and query: \n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL. Nilvadipine is a non-steroidal anti-inflammatory drug (NSAID) that is used to treat CADASIL.']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling. \n",
      "\n",
      "evidence 0: Using both wild type mice and transgenic mice carrying the human mutant Notch3 gene (CADASIL mice), we have recently characterized the pathological features of CADASIL and determined the therapeutic efficacy of two hematopoietic growth factors, stem cell factor (SCF) and granulocyte colony-stimulating factor (G-CSF) in CADASIL. Our findings have revealed novel pathological changes in the endothelium of cerebral capillaries and in the neural stem cells (NSCs). We have also observed the impairment of cognitive function in CADASIL mice. \n",
      "evidence 1: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available. \n",
      "evidence 2: Today, CADASIL is known to every neurologist, but the disease has not yet revealed all its secrets. A lot of effort is still needed to understand the intimate mechanisms of the disease and the most efficient targets or approaches for the development of efficient therapeutics. The history of CADASIL will be further enriched by multiple ongoing research projects worldwide, at clinical and preclinical level, and will continue to enlighten research in the field of cerebral small vessel disorders. \n",
      "query: what are the potential therapeutic targets for cadasil? \n",
      "Summarize given evidences and query: \n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling. \n",
      "\n",
      "evidence 0: Syk inhibition was found to stabilize microtubules and potentiate paclitaxel activity in cellular models of taxane-resistant ovarian cancers. We further studied the effects of Syk inhibition on paclitaxel activity in Syk(+) ovarian cancer cell models and in variants selected for taxane resistance. Syk inhibition was accomplished using RNAi and by exposure to the small molecule competitive inhibitor R406, the active metabolite of fostamatinib. \n",
      "evidence 1: Therefore, inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients. \n",
      "evidence 2: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available. \n",
      "query: would SYK inhibition be potential treatment for cadasil? \n",
      "Summarize given evidences and query: \n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMCs) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling. Inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients.']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling. \n",
      "\n",
      "evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF-like domains. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. \n",
      "evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 \n",
      "evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC. ...maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC. \n",
      "query: What is the cause of the CADASIL? \n",
      "Summarize given evidences and query: \n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['CADASIL is a hereditary vascular dementia, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC....maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC.']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling. \n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
      "evidence 1: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
      "evidence 2: Niclosamide and derivatives - potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal anti-inflammatory compound that inhibits goblet cell degranulation and suppresses asthma phenotype (4, 27). ...This result suggests an antiinflammatory effect of niclosamide, which corresponds well to inhibition of allergic lung inflammation by the TMEM16A inhibitor benzbromarone (3, 31). \n",
      "evidence 3: The key findings of this study are that (a) Cl- fluxes mediated by the Ca2+-gated channel TMEM16A are a crucial determinant of pericyte tone; (b) TMEM16A is activated during ischemia and evokes a long-lasting pericytemediated capillary constriction that reduces CBF and favors neutrophil and platelet stalling; (c) genetic analysis suggests that increased TMEM16A expression is associated with poor recovery after ischemic stroke (and the genetic proxy \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: \n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['Question: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling.']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling. \n",
      "\n",
      "evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy ... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with... \n",
      "evidence 1: Differences in proliferation rate between CADASIL and control vascular smooth muscle cells are ... (J Cell Mol Med 2018). We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. ...showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic... \n",
      "evidence 2: The Role of Vascular Smooth Muscle Cells in Arterial Remodeling: Focus on Calcification ... (Int J Mol Sci 2019). Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused... \n",
      "evidence 3: Spleen Tyrosine Kinase (SYK) in the Progression of Peritoneal Fibrosis Through Activation of the ... (Med Sci Monit 2019). Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase... \n",
      "evidence 4: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage... \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: \n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of the TGFb1 signaling pathway.']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGFb1 signaling. \n",
      "\n",
      "evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy ... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with... \n",
      "evidence 1: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage... \n",
      "evidence 2: Niclosamide repurposed for the treatment of inflammatory airway disease (JCI Insight 2019). Niclosamide and derivatives - potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal antiinflammatory compound that inhibits goblet... \n",
      "evidence 3: The Ca2+-gated channel TMEM16A amplifies capillary pericyte contraction and reduces ... (J Clin Invest. 2022). The key findings of this study are that (a) Cl- fluxes mediated by the Ca2+-gated channel TMEM16A are a crucial determinant of pericyte tone \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "Summarize given evidences and query: \n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of the TGFb1 signaling pathway.']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(f'{output_file_name}.txt', \"w\") as results: # mujeen_\n",
    "    results.write(f'<< Model: {model_name} >>\\n\\n')\n",
    "    \n",
    "    for prompt in prompts:\n",
    "        print('< PROMPT >\\n')\n",
    "        print(prompt)\n",
    "        results.write('< PROMPT >\\n\\n')\n",
    "        results.write(prompt)\n",
    "\n",
    "        inputs = tokenizer(prompt, return_tensors=\"pt\", max_length=1024, truncation=True)\n",
    "        inputs.to(device)\n",
    "        outputs = model.generate(**inputs, max_length=1024, min_length=100, num_beams=5)\n",
    "        generated = tokenizer.batch_decode(outputs, skip_special_tokens=True)\n",
    "        \n",
    "        # output_dic['output'].append({'prompt': prompt, 'generated': *generated})\n",
    "\n",
    "        print('\\n\\n< GENERATED >\\n')\n",
    "        print(generated)\n",
    "        print('\\n###############################################################################################################################################################\\n')\n",
    "        results.write('\\n\\n< GENERATED >\\n\\n')\n",
    "        results.write(*generated)\n",
    "        results.write('\\n\\n###############################################################################################################################################################\\n')\n",
    "        \n",
    "results.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biomedlm",
   "language": "python",
   "name": "biomedlm"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "vscode": {
   "interpreter": {
    "hash": "ce5cb1d3583c4bc4c26ba2d9d1424221025f98623b08f3ff314497f7b1d6b7f8"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
