{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "os.environ[\"CUDA_DEVICE_ORDER\"]=\"PCI_BUS_ID\"\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"]=\"0, 1, 2, 3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "256\n"
     ]
    }
   ],
   "source": [
    "print(os.system('which transformers'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_name = \"BioGPT-Large\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cuda\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "BioGptForCausalLM(\n",
       "  (biogpt): BioGptModel(\n",
       "    (embed_tokens): Embedding(57717, 1600, padding_idx=1)\n",
       "    (embed_positions): BioGptLearnedPositionalEmbedding(2050, 1600)\n",
       "    (layers): ModuleList(\n",
       "      (0): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (1): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (2): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (3): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (4): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (5): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (6): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (7): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (8): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (9): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (10): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (11): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (12): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (13): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (14): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (15): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (16): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (17): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (18): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (19): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (20): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (21): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (22): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (23): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (24): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (25): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (26): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (27): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (28): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (29): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (30): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (31): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (32): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (33): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (34): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (35): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (36): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (37): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (38): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (39): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (40): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (41): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (42): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (43): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (44): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (45): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (46): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "      (47): BioGptDecoderLayer(\n",
       "        (self_attn): BioGptAttention(\n",
       "          (k_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (v_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (q_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "          (out_proj): Linear(in_features=1600, out_features=1600, bias=True)\n",
       "        )\n",
       "        (activation_fn): GELUActivation()\n",
       "        (self_attn_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "        (fc1): Linear(in_features=1600, out_features=6400, bias=True)\n",
       "        (fc2): Linear(in_features=6400, out_features=1600, bias=True)\n",
       "        (final_layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "      )\n",
       "    )\n",
       "    (layer_norm): LayerNorm((1600,), eps=1e-05, elementwise_affine=True)\n",
       "  )\n",
       "  (output_projection): Linear(in_features=1600, out_features=57717, bias=False)\n",
       ")"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from transformers import AutoTokenizer, AutoModelForCausalLM\n",
    "import torch\n",
    "\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "print(device)\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"microsoft/BioGPT-Large\")\n",
    "\n",
    "model = AutoModelForCausalLM.from_pretrained(\"microsoft/BioGPT-Large\")\n",
    "model.to(device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "803380b6f5e44e3d8066ebc507e2d25f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading:   0%|          | 0.00/1.44k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2d0bb70bb68047eeb34c1877c5dd4666",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading:   0%|          | 0.00/50.8k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b87cee16bafd414e8259713844428433",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading:   0%|          | 0.00/9.45G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "59165c2ed8c3415394a59090fe54e0d7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading:   0%|          | 0.00/1.95G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d25d1a0866074d86a25a6a0a985176eb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading:   0%|          | 0.00/2.54k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ece7f2e9c51a43a9aae90c544be77191",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading:   0%|          | 0.00/792k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dc664ec0fa1b490cabc768cbac4cb850",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading:   0%|          | 0.00/2.42M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "de7b27d3aa7e45588cfc1194249b5a21",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading:   0%|          | 0.00/2.20k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# ## https://github.com/dmis-lab/HypoGPT/blob/main/humanfeedback/flan-t5.py\n",
    "\n",
    "# from transformers import AutoConfig, AutoModelForSeq2SeqLM, AutoTokenizer\n",
    "# import json\n",
    "\n",
    "# # model_name = 'google/flan-t5-small'\n",
    "# # model_name = 'google/flan-t5-base'\n",
    "# # model_name = 'google/flan-t5-small'\n",
    "# model_name = 'google/flan-t5-xl'\n",
    "# # model_name = 'google/flan-t5-xxl'\n",
    "# config = AutoConfig.from_pretrained(model_name)\n",
    "# config.n_positions = 2048\n",
    "# model = AutoModelForSeq2SeqLM.from_pretrained(model_name, config=config)\n",
    "# tokenizer = AutoTokenizer.from_pretrained(model_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 교수님께서 주신 예시\n",
    "\n",
    "item1 = \"\"\"Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms.[1]\"\"\"\n",
    "item2 = \"\"\"Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18[2]\"\"\"\n",
    "item3 = \"\"\"CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC. ...maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC.[3]\"\"\"\n",
    "\n",
    "item4 = \"There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36[1]\"\n",
    "item5 = \"We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease.[2]\"\n",
    "item6 = \"Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,...[3]\"\n",
    "item7 = \"Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK.[4]\"\n",
    "item8 = \"Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results,[5]\"\n",
    "item9 = \"Niclosamide and derivatives - potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal anti-inflammatory compound that inhibits goblet cell degranulation and suppresses asthma phenotype (4, 27). ...This result suggests an antiinflammatory effect of niclosamide, which corresponds well to inhibition of allergic lung inflammation by the TMEM16A inhibitor benzbromarone (3, 31).[6]\"\n",
    "item10 = \"The key findings of this study are that (a) Cl- fluxes mediated by the Ca2+gated channel TMEM16A are a crucial determinant of pericyte tone; (b) TMEM16A is activated during ischemia and evokes a long-lasting pericytemediated capillary constriction that reduces CBF and favors neutrophil and platelet stalling; (c) genetic analysis suggests that increased TMEM16A expression is associated with poor recovery after ischemic stroke (and the genetic proxy[7]\"\n",
    "\n",
    "item11 = \"Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy ... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with ...[1]\"\n",
    "item12 = \"Differences in proliferation rate between CADASIL and control vascular smooth muscle cells are ... (J Cell Mol Med 2018). We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. ... showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic...[2]\"\n",
    "item13 = \"The Role of Vascular Smooth Muscle Cells in Arterial Remodeling: Focus on Calcification ... (Int J Mol Sci 2019). Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused ...[3]\"\n",
    "item14 = \"Spleen Tyrosine Kinase (SYK) in the Progression of Peritoneal Fibrosis Through Activation of the ... (Med Sci Monit 2019). Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase ...[4]\"\n",
    "item15 = \"NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage ...[5]\"\n",
    "item16 = \"Niclosamide repurposed for the treatment of inflammatory airway disease (JCI Insight 2019). Niclosamide and derivatives - potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal antiinflammatory compound that inhibits goblet ...[6]\"\n",
    "item17 = \"The Ca2+-gated channel TMEM16A amplifies capillary pericyte contraction and reduces ... (J Clin Invest. 2022). The key findings of this study are that (a) Cl- fluxes mediated by the Ca2+-gated channel TMEM16A are a crucial determinant of pericyte tone[7]\"\n",
    "\n",
    "gt1 = \"CADASIL is a rare hereditary cerebrovascular disease caused by a NOTCH3 mutation[1][2]. NOTCH3 is particularly important for the maturation of vascular smooth muscle cells (VSMCs). A decreased VSMC is a pathological hallmark for CADASIL[3].\"\n",
    "gt2 = \"Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated[1] through the increase of TGFb1 signaling[2]. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells[3]. Syk increases fibrosis through activation of TGFb1 signaling[4].\"\n",
    "gt3 = \"TMEM16A is a therapeutic target for CADASIL. Ischemia in CADASIL activates TMEM16A, Ca2+ gated Cl- channel, which regulates pericyte tone and induces capillary constriction[1]. Potent inhibitors of TMEM16A, niclosamide has antiinflammtory effect[2]. Niclosamide can suppress enhance inflammation in CADASIL[3].\"\n",
    "gt4 = \"Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated[1] through the increase of TGFb1 signaling[2] . The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells[3] . Syk increases fibrosis through activation of TGFb1 signaling[4].\"\n",
    "gt5 = \"TMEM16A is a therapeutic target for CADASIL. Ischemia in CADASIL activates TMEM16A, Ca2+ gated Cl- channel, which regulates pericyte tone and induces capillary constriction[1]. Potent inhibitors of TMEM16A, niclosamide has antiinflammtory effect[2]. Niclosamide can suppress enhance inflammation in CADASIL[3].\"\n",
    "\n",
    "q1 = \"What is the cause of the CADASIL?\"\n",
    "q2 = \"What is the therapeutic target for CADASIL?\"\n",
    "q3 = \"What is the therapeutic target for CADASIL?\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 현재님께서 retrieve하신 evidences\n",
    "\n",
    "rt_item1 = \"CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship.[1]\"\n",
    "rt_item2 = \"Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing.[2]\"\n",
    "rt_item3 = \"Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL.[3]\"\n",
    "\n",
    "rt_item4 = \"Using both wild type mice and transgenic mice carrying the human mutant Notch3 gene (CADASIL mice), we have recently characterized the pathological features of CADASIL and determined the therapeutic efficacy of two hematopoietic growth factors, stem cell factor (SCF) and granulocyte colony-stimulating factor (G-CSF) in CADASIL. Our findings have revealed novel pathological changes in the endothelium of cerebral capillaries and in the neural stem cells (NSCs). We have also observed the impairment of cognitive function in CADASIL mice.[1]\"\n",
    "rt_item5 = \"Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.[2]\"\n",
    "rt_item6 = \"Today, CADASIL is known to every neurologist, but the disease has not yet revealed all its secrets. A lot of effort is still needed to understand the intimate mechanisms of the disease and the most efficient targets or approaches for the development of efficient therapeutics. The history of CADASIL will be further enriched by multiple ongoing research projects worldwide, at clinical and preclinical level, and will continue to enlighten research in the field of cerebral small vessel disorders.[3]\"\n",
    "\n",
    "rt_item7 = \"Syk inhibition was found to stabilize microtubules and potentiate paclitaxel activity in cellular models of taxane-resistant ovarian cancers. We further studied the effects of Syk inhibition on paclitaxel activity in Syk(+) ovarian cancer cell models and in variants selected for taxane resistance. Syk inhibition was accomplished using RNAi and by exposure to the small molecule competitive inhibitor R406, the active metabolite of fostamatinib.[1]\"\n",
    "rt_item8 = \"Therefore, inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients.[2]\"\n",
    "rt_item9 = \"Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.[3]\"\n",
    "\n",
    "\n",
    "rt_q1 = \"can nilvadipine help alleviate cadasil symptoms?\"\n",
    "rt_q2 = \"what are the potential therapeutic targets for cadasil?\"\n",
    "rt_q3 = \"would SYK inhibition be potential treatment for cadasil?\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## dic['data'] = list of dictionaries. each dic has fields: \"query\", \"evidences\"(itself a list of strings), \"ground_truth\"\n",
    "## pre-compute data dictionary file. yea? yep. \n",
    "## see how the code handles cases without ground truth -> from same example dic, just doesn't take the ground truth. \n",
    "## make three with complete examples, three with retrieved evidences, \n",
    "## or make two json files one with the three complete examples and the other with retrieved evidences. \n",
    "\n",
    "import json\n",
    "\n",
    "context1 = {\"query\": q1, \"evidences\": [item1, item2, item3], \"ground_truth\": gt1}\n",
    "\n",
    "context2 = {\"query\": q2, \"evidences\": [item4, item5, item6, item7, item8], \"ground_truth\": gt2}\n",
    "context3 = {\"query\": q2, \"evidences\": [item4, item8, item9, item10], \"ground_truth\": gt3}\n",
    "\n",
    "context4 = {\"query\": q3, \"evidences\": [item11, item12, item13, item14, item15], \"ground_truth\": gt4}\n",
    "context5 = {\"query\": q3, \"evidences\": [item11, item15, item16, item17], \"ground_truth\": gt5}\n",
    "\n",
    "query1 = {\"query\": rt_q1, \"evidences\": [rt_item1, rt_item2, rt_item3]}\n",
    "query2 = {\"query\": rt_q2, \"evidences\": [rt_item4, rt_item5, rt_item6]}\n",
    "query3 = {\"query\": rt_q3, \"evidences\": [rt_item7, rt_item8, rt_item9]}\n",
    "\n",
    "dic = {'contexts': [context1, context2, context3, context4, context5], 'queries': [query1, query2, query3]}\n",
    "\n",
    "with open(\"examples_20230207.json\", \"w\") as outfile:\n",
    "    json.dump(dic, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('examples_20230207.json') as f:\n",
    "    loaded_dic = json.load(f)\n",
    "    contexts = loaded_dic['contexts']\n",
    "    queries = loaded_dic['queries']\n",
    "    \n",
    "    \n",
    "## any combination of contexts[i] and queries[i] possible? -> well contexts can be query but not the other way around\n",
    "# also, context 23 and 45 are v similar\n",
    "\n",
    "args = [(contexts[0], contexts[1]), (contexts[0], contexts[2]), (contexts[0], contexts[3]), (contexts[0], contexts[4]),\\\n",
    "        (contexts[1], contexts[0]), (contexts[2], contexts[0]), (contexts[3], contexts[0]), (contexts[4], contexts[0]),\\\n",
    "        (contexts[1], contexts[2]), (contexts[2], contexts[1]),\\\n",
    "        (contexts[3], contexts[4]), (contexts[4], contexts[3]),\\\n",
    "        (contexts[0], queries[0]), (contexts[1], queries[0]), (contexts[2], queries[0]),\\\n",
    "        (contexts[0], queries[1]), (contexts[1], queries[1]), (contexts[2], queries[1]),\\\n",
    "        (contexts[0], queries[2]), (contexts[1], queries[2]), (contexts[2], queries[2])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def make_prompt(example, query, zeroshot=False): # both args are of type either context dic or query dic -> {\"query\": q3, \"evidences\": [item11, item15, item16, item17], \"ground_truth\": gt5}\n",
    "    \n",
    "    example_part = \"\"\n",
    "    if not zeroshot:\n",
    "        for i, ev in enumerate(example['evidences']):\n",
    "            example_part += f\"evidence {i}: {ev} \\n\"\n",
    "\n",
    "        example_part += f\"query: {example['query']} \\n\"\n",
    "        example_part += \"summarize given evidences and query:\"\n",
    "        example_part += f\"{example['ground_truth']} \\n\\n\"\n",
    "    \n",
    "    query_part = \"\"\n",
    "    for i, ev in enumerate(query['evidences']):\n",
    "        query_part += f\"evidence {i}: {ev} \\n\"\n",
    "    \n",
    "    query_part += f\"query: {query['query']} \\n\"\n",
    "    query_part += \"summarize given evidences and query:\"\n",
    "    \n",
    "    prompt =  example_part + query_part\n",
    "    return prompt\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'BioGptTokenizer' object has no attribute 'to'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[0;32mIn [11]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mtokenizer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m(device)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'BioGptTokenizer' object has no attribute 'to'"
     ]
    }
   ],
   "source": [
    "tokenizer.to(device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "item1 = '''Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients. All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms.'''\n",
    "item2 = '''Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18'''\n",
    "item3 = '''CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of vSMC . maturation of vSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of vSMC.'''\n",
    "\n",
    "item3 = \"There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \"\n",
    "item4 = \"We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \"\n",
    "item5 = \"Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large\"\n",
    "item6 = \"Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK.\"\n",
    "item7 = \"Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-κB. A positive regulation between NOTCH and NF-κB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \"\n",
    "\n",
    "gt = \"Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGF-b1 signaling.\"\n",
    "\n",
    "set1 = {\"query\": 'What is the cause of the CADASIL?', \"evidences\": [item1, item2, item3]}\n",
    "set2 = {\"query\": 'What is the therapeutic target for CADASIL?', \"evidences\": [item3, item4, item5, item6, item7], \"ground_truth\": gt}\n",
    "\n",
    "args = (set2, set1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36  \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease.  \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-κB. A positive regulation between NOTCH and NF-κB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results,  \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGF-b1 signaling. \n",
      "\n",
      "evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients. All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. \n",
      "evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 \n",
      "evidence 2: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36  \n",
      "query: What is the cause of the CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive, 36 evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1 / Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-κB. A positive regulation between NOTCH and NF-κB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, query: What is the therapeutic target for CADASIL? summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGF-b1 signaling. evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients. All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 evidence 2: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive, 36 query: What is the cause of the CADASIL? summarize given evidences and query: The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormality.Evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients. All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. evidence 1: Linkage studies in other families enabled further refinement of this genetic interval6, 17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 evidence 2: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients. All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains.24 evidence 3: Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signaling play a key role in the pathogenesis and progression of chronic cardiovascular diseases. Support for our']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "with open(f\"generation_results_mujeen_{model_name.replace('/', '-')}_one-shot.txt\", \"w\") as results:\n",
    "    results.write(f'<< Model: {model_name} >>\\n\\n')\n",
    "    # results.write()\n",
    "\n",
    "    # for i in range(len(args)):\n",
    "    for i in range(1):\n",
    "        example, query = args #[i]\n",
    "        prompt = make_prompt(example, query)\n",
    "        \n",
    "        print('< PROMPT >\\n')\n",
    "        print(prompt)\n",
    "        results.write('< PROMPT >\\n\\n')\n",
    "        results.write(prompt)\n",
    "        \n",
    "        # inputs = tokenizer(prompt, return_tensors=\"pt\", max_length=2048, truncation=True)\n",
    "        inputs = tokenizer(prompt, return_tensors=\"pt\", max_length=1024, truncation=True)\n",
    "        inputs.to(device)\n",
    "        outputs = model.generate(**inputs, max_length=1024)\n",
    "        generated = tokenizer.batch_decode(outputs, skip_special_tokens=True)\n",
    "        \n",
    "        print('\\n\\n< GENERATED >\\n')\n",
    "        print(generated)\n",
    "        print('\\n###############################################################################################################################################################\\n')\n",
    "        results.write('\\n\\n< GENERATED >\\n\\n')\n",
    "        results.write(*generated)\n",
    "        results.write('\\n\\n###############################################################################################################################################################\\n')\n",
    "        \n",
    "results.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36[1] \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease.[2] \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,...[3] \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK.[4] \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results,[5] \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive, 36 [1] evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. [2] evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... [3] evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules [25].... bidirectional [33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1 / Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. [4] evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, [5] query: What is the therapeutic target for CADASIL? summarize given evidences and query: We believe that the therapeutic target for CADASIL is the NOTCH3 receptor itself. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > In conclusion, we have shown that the CADASIL-associated mutations in the NOTCH3 gene lead to increased expression of TGFb3 and that this is associated with the activation of the SYK / TGF-b1 / Smad3 signaling pathway. We also showed that the activation of NF-kB is a key event in the progression of peritoneal fibrosis. Our results suggest that the NOTCH3 receptor itself is a potential therapeutic target for CADASIL. < / FREETEXT > < / PARAGRAPH > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36[1] \n",
      "evidence 1: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results,[5] \n",
      "evidence 2: Niclosamide and derivatives - potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal anti-inflammatory compound that inhibits goblet cell degranulation and suppresses asthma phenotype (4, 27). ...This result suggests an antiinflammatory effect of niclosamide, which corresponds well to inhibition of allergic lung inflammation by the TMEM16A inhibitor benzbromarone (3, 31).[6] \n",
      "evidence 3: The key findings of this study are that (a) Cl- fluxes mediated by the Ca2+gated channel TMEM16A are a crucial determinant of pericyte tone; (b) TMEM16A is activated during ischemia and evokes a long-lasting pericytemediated capillary constriction that reduces CBF and favors neutrophil and platelet stalling; (c) genetic analysis suggests that increased TMEM16A expression is associated with poor recovery after ischemic stroke (and the genetic proxy[7] \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive, 36 [1] evidence 1: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, [5] evidence 2: Niclosamide and derivatives - potent inhibitors of TMEM16A / F and Ca2 + signaling. NFA is a nonsteroidal anti-inflammatory compound that inhibits goblet cell degranulation and suppresses asthma phenotype (4, 27).... This result suggests an antiinflammatory effect of niclosamide, which corresponds well to inhibition of allergic lung inflammation by the TMEM16A inhibitor benzbromarone (3, 31). [6] evidence 3: The key findings of this study are that (a) Cl- fluxes mediated by the Ca2 + gated channel TMEM16A are a crucial determinant of pericyte tone; (b) TMEM16A is activated during ischemia and evokes a long-lasting pericytemediated capillary constriction that reduces CBF and favors neutrophil and platelet stalling; (c) genetic analysis suggests that increased TMEM16A expression is associated with poor recovery after ischemic stroke (and the genetic proxy [7] query: What is the therapeutic target for CADASIL? summarize given evidences and query: What is the therapeutic target for CADASIL? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > In conclusion, our study provides evidence that TMEM16A is a key regulator of pericyte tone and that its activation is a key event in the pathogenesis of CADASIL. Our results suggest that TMEM16A is a potential therapeutic target for CADASIL and possibly other diseases associated with pericyte dysfunction. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > This work was supported by the French National Research Agency (ANR), the French Ministry of Research (ANR-1 0-LABX-6 2-IBEID), the French Ministry of Health (PHRC 2 0 1 4 – 2 0 1 6), the French Ministry of Education and Research (FRM), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation Leducq (1 1CVD0 4), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation pour la Recherche Médicale (FRM), the Fondation Leducq (1 1CVD0 4), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy ... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with ...[1] \n",
      "evidence 1: Differences in proliferation rate between CADASIL and control vascular smooth muscle cells are ... (J Cell Mol Med 2018). We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. ... showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic...[2] \n",
      "evidence 2: The Role of Vascular Smooth Muscle Cells in Arterial Remodeling: Focus on Calcification ... (Int J Mol Sci 2019). Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused ...[3] \n",
      "evidence 3: Spleen Tyrosine Kinase (SYK) in the Progression of Peritoneal Fibrosis Through Activation of the ... (Med Sci Monit 2019). Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase ...[4] \n",
      "evidence 4: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage ...[5] \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with... [1] evidence 1: Differences in proliferation rate between CADASIL and control vascular smooth muscle cells are... (J Cell Mol Med 2018). We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL.... showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic... [2] evidence 2: The Role of Vascular Smooth Muscle Cells in Arterial Remodeling: Focus on Calcification... (Int J Mol Sci 2019). Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused... [3] evidence 3: Spleen Tyrosine Kinase (SYK) in the Progression of Peritoneal Fibrosis Through Activation of the... (Med Sci Monit 2019). Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase... [4] evidence 4: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage... [5] query: What is the therapeutic target for CADASIL? summarize given evidences and query: What is the therapeutic target for CADASIL? < / FREETEXT > < / ABSTRACT > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy ... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with ...[1] \n",
      "evidence 1: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage ...[5] \n",
      "evidence 2: Niclosamide repurposed for the treatment of inflammatory airway disease (JCI Insight 2019). Niclosamide and derivatives - potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal antiinflammatory compound that inhibits goblet ...[6] \n",
      "evidence 3: The Ca2+-gated channel TMEM16A amplifies capillary pericyte contraction and reduces ... (J Clin Invest. 2022). The key findings of this study are that (a) Cl- fluxes mediated by the Ca2+-gated channel TMEM16A are a crucial determinant of pericyte tone[7] \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with... [1] evidence 1: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage... [5] evidence 2: Niclosamide repurposed for the treatment of inflammatory airway disease (JCI Insight 2019). Niclosamide and derivatives - potent inhibitors of TMEM16A / F and Ca2 + signaling. NFA is a nonsteroidal antiinflammatory compound that inhibits goblet... [6] evidence 3: The Ca2 + -gated channel TMEM16A amplifies capillary pericyte contraction and reduces... (J Clin Invest. 2022). The key findings of this study are that (a) Cl- fluxes mediated by the Ca2 + -gated channel TMEM16A are a crucial determinant of pericyte tone [7] query: What is the therapeutic target for CADASIL? summarize given evidences and query: What is the therapeutic target for CADASIL? < / FREETEXT > < / ABSTRACT > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms.[1] \n",
      "evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18[2] \n",
      "evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC. ...maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC.[3] \n",
      "query: What is the cause of the CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. [1] evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 [2] evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC.... maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC. [3] query: What is the cause of the CADASIL? summarize given evidences and query: What is the cause of the CADASIL? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The identification of the Notch3 gene as the defective gene in CADASIL patients has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The identification of the Notch3 gene as the defective gene in CADASIL patients has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The authors thank the patients and their families for their participation in this study. This work was supported by grants from the Italian Ministry of Health (Ricerca Finalizzata) and the Italian Ministry of Education, University and Research (PRIN 2 0 0 0). < / FREETEXT > < / PARAGRAPH > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms.[1] \n",
      "evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18[2] \n",
      "evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC. ...maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC.[3] \n",
      "query: What is the cause of the CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. [1] evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 [2] evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC.... maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC. [3] query: What is the cause of the CADASIL? summarize given evidences and query: What is the cause of the CADASIL? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The identification of the Notch3 gene as the defective gene in CADASIL patients has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The identification of the Notch3 gene as the defective gene in CADASIL patients has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The authors thank the patients and their families for their participation in this study. This work was supported by grants from the Italian Ministry of Health (Ricerca Finalizzata) and the Italian Ministry of Education, University and Research (PRIN 2 0 0 0). < / FREETEXT > < / PARAGRAPH > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms.[1] \n",
      "evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18[2] \n",
      "evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC. ...maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC.[3] \n",
      "query: What is the cause of the CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. [1] evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 [2] evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC.... maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC. [3] query: What is the cause of the CADASIL? summarize given evidences and query: What is the cause of the CADASIL? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The identification of the Notch3 gene as the defective gene in CADASIL patients has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The identification of the Notch3 gene as the defective gene in CADASIL patients has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The authors thank the patients and their families for their participation in this study. This work was supported by grants from the Italian Ministry of Health (Ricerca Finalizzata) and the Italian Ministry of Education, University and Research (PRIN 2 0 0 0). < / FREETEXT > < / PARAGRAPH > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms.[1] \n",
      "evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18[2] \n",
      "evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC. ...maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC.[3] \n",
      "query: What is the cause of the CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. [1] evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 [2] evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC.... maturation of VSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of VSMC. [3] query: What is the cause of the CADASIL? summarize given evidences and query: What is the cause of the CADASIL? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The identification of the Notch3 gene as the defective gene in CADASIL patients has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The identification of the Notch3 gene as the defective gene in CADASIL patients has been a major step forward in the understanding of the pathogenesis of this disease. The identification of the Notch3 gene has been a major step forward in the understanding of the pathogenesis of this disease. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The authors thank the patients and their families for their participation in this study. This work was supported by grants from the Italian Ministry of Health (Ricerca Finalizzata) and the Italian Ministry of Education, University and Research (PRIN 2 0 0 0). < / FREETEXT > < / PARAGRAPH > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36[1] \n",
      "evidence 1: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results,[5] \n",
      "evidence 2: Niclosamide and derivatives - potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal anti-inflammatory compound that inhibits goblet cell degranulation and suppresses asthma phenotype (4, 27). ...This result suggests an antiinflammatory effect of niclosamide, which corresponds well to inhibition of allergic lung inflammation by the TMEM16A inhibitor benzbromarone (3, 31).[6] \n",
      "evidence 3: The key findings of this study are that (a) Cl- fluxes mediated by the Ca2+gated channel TMEM16A are a crucial determinant of pericyte tone; (b) TMEM16A is activated during ischemia and evokes a long-lasting pericytemediated capillary constriction that reduces CBF and favors neutrophil and platelet stalling; (c) genetic analysis suggests that increased TMEM16A expression is associated with poor recovery after ischemic stroke (and the genetic proxy[7] \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive, 36 [1] evidence 1: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, [5] evidence 2: Niclosamide and derivatives - potent inhibitors of TMEM16A / F and Ca2 + signaling. NFA is a nonsteroidal anti-inflammatory compound that inhibits goblet cell degranulation and suppresses asthma phenotype (4, 27).... This result suggests an antiinflammatory effect of niclosamide, which corresponds well to inhibition of allergic lung inflammation by the TMEM16A inhibitor benzbromarone (3, 31). [6] evidence 3: The key findings of this study are that (a) Cl- fluxes mediated by the Ca2 + gated channel TMEM16A are a crucial determinant of pericyte tone; (b) TMEM16A is activated during ischemia and evokes a long-lasting pericytemediated capillary constriction that reduces CBF and favors neutrophil and platelet stalling; (c) genetic analysis suggests that increased TMEM16A expression is associated with poor recovery after ischemic stroke (and the genetic proxy [7] query: What is the therapeutic target for CADASIL? summarize given evidences and query: What is the therapeutic target for CADASIL? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > In conclusion, our study provides evidence that TMEM16A is a key regulator of pericyte tone and that its activation is a key event in the pathogenesis of CADASIL. Our results suggest that TMEM16A is a potential therapeutic target for CADASIL and possibly other diseases associated with pericyte dysfunction. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > This work was supported by the French National Research Agency (ANR), the French Ministry of Research (ANR-1 0-LABX-6 2-IBEID), the French Ministry of Health (PHRC 2 0 1 4 – 2 0 1 6), the French Ministry of Education and Research (FRM), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation Leducq (1 1CVD0 4), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation pour la Recherche Médicale (FRM), the Fondation Leducq (1 1CVD0 4), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4 1 2 0 0 3 7 7), the Fondation Leducq (1 1CVD0 4), the Fondation pour la Recherche Médicale (FRM), the Fondation ARC pour la Recherche sur le Cancer (PJA 2 0 1 4']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36[1] \n",
      "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease.[2] \n",
      "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,...[3] \n",
      "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules[25]. ...bidirectional[33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK.[4] \n",
      "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results,[5] \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive, 36 [1] evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. [2] evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large,... [3] evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules [25].... bidirectional [33]. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1 / Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. [4] evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-kB. A positive regulation between NOTCH and NF-kB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, [5] query: What is the therapeutic target for CADASIL? summarize given evidences and query: We believe that the therapeutic target for CADASIL is the NOTCH3 receptor itself. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > In conclusion, we have shown that the CADASIL-associated mutations in the NOTCH3 gene lead to increased expression of TGFb3 and that this is associated with the activation of the SYK / TGF-b1 / Smad3 signaling pathway. We also showed that the activation of NF-kB is a key event in the progression of peritoneal fibrosis. Our results suggest that the NOTCH3 receptor itself is a potential therapeutic target for CADASIL. < / FREETEXT > < / PARAGRAPH > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy ... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with ...[1] \n",
      "evidence 1: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage ...[5] \n",
      "evidence 2: Niclosamide repurposed for the treatment of inflammatory airway disease (JCI Insight 2019). Niclosamide and derivatives - potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal antiinflammatory compound that inhibits goblet ...[6] \n",
      "evidence 3: The Ca2+-gated channel TMEM16A amplifies capillary pericyte contraction and reduces ... (J Clin Invest. 2022). The key findings of this study are that (a) Cl- fluxes mediated by the Ca2+-gated channel TMEM16A are a crucial determinant of pericyte tone[7] \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with... [1] evidence 1: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage... [5] evidence 2: Niclosamide repurposed for the treatment of inflammatory airway disease (JCI Insight 2019). Niclosamide and derivatives - potent inhibitors of TMEM16A / F and Ca2 + signaling. NFA is a nonsteroidal antiinflammatory compound that inhibits goblet... [6] evidence 3: The Ca2 + -gated channel TMEM16A amplifies capillary pericyte contraction and reduces... (J Clin Invest. 2022). The key findings of this study are that (a) Cl- fluxes mediated by the Ca2 + -gated channel TMEM16A are a crucial determinant of pericyte tone [7] query: What is the therapeutic target for CADASIL? summarize given evidences and query: What is the therapeutic target for CADASIL? < / FREETEXT > < / ABSTRACT > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy ... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with ...[1] \n",
      "evidence 1: Differences in proliferation rate between CADASIL and control vascular smooth muscle cells are ... (J Cell Mol Med 2018). We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. ... showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic...[2] \n",
      "evidence 2: The Role of Vascular Smooth Muscle Cells in Arterial Remodeling: Focus on Calcification ... (Int J Mol Sci 2019). Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused ...[3] \n",
      "evidence 3: Spleen Tyrosine Kinase (SYK) in the Progression of Peritoneal Fibrosis Through Activation of the ... (Med Sci Monit 2019). Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase ...[4] \n",
      "evidence 4: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage ...[5] \n",
      "query: What is the therapeutic target for CADASIL? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy... (Stroke 2018). There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with... [1] evidence 1: Differences in proliferation rate between CADASIL and control vascular smooth muscle cells are... (J Cell Mol Med 2018). We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL.... showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic... [2] evidence 2: The Role of Vascular Smooth Muscle Cells in Arterial Remodeling: Focus on Calcification... (Int J Mol Sci 2019). Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused... [3] evidence 3: Spleen Tyrosine Kinase (SYK) in the Progression of Peritoneal Fibrosis Through Activation of the... (Med Sci Monit 2019). Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase... [4] evidence 4: NOTCH3 signaling is essential for NF-kB activation in TLRactivated macrophages (Sci Rep 2020). Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage... [5] query: What is the therapeutic target for CADASIL? summarize given evidences and query: What is the therapeutic target for CADASIL? < / FREETEXT > < / ABSTRACT > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship.[1] \n",
      "evidence 1: Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing.[2] \n",
      "evidence 2: Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL.[3] \n",
      "query: can nilvadipine help alleviate cadasil symptoms? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. [1] evidence 1: Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing. [2] evidence 2: Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL. [3] query: can nilvadipine help alleviate cadasil symptoms? summarize given evidences and query: nilvadipine has been shown to be effective in the treatment of cadasil symptoms. < / FREETEXT > < / ABSTRACT > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship.[1] \n",
      "evidence 1: Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing.[2] \n",
      "evidence 2: Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL.[3] \n",
      "query: can nilvadipine help alleviate cadasil symptoms? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. [1] evidence 1: Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing. [2] evidence 2: Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL. [3] query: can nilvadipine help alleviate cadasil symptoms? summarize given evidences and query: nilvadipine has been shown to be effective in the treatment of cadasil symptoms. < / FREETEXT > < / ABSTRACT > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship.[1] \n",
      "evidence 1: Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing.[2] \n",
      "evidence 2: Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL.[3] \n",
      "query: can nilvadipine help alleviate cadasil symptoms? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. [1] evidence 1: Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing. [2] evidence 2: Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL. [3] query: can nilvadipine help alleviate cadasil symptoms? summarize given evidences and query: nilvadipine has been shown to be effective in the treatment of cadasil symptoms. < / FREETEXT > < / ABSTRACT > ▃']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Using both wild type mice and transgenic mice carrying the human mutant Notch3 gene (CADASIL mice), we have recently characterized the pathological features of CADASIL and determined the therapeutic efficacy of two hematopoietic growth factors, stem cell factor (SCF) and granulocyte colony-stimulating factor (G-CSF) in CADASIL. Our findings have revealed novel pathological changes in the endothelium of cerebral capillaries and in the neural stem cells (NSCs). We have also observed the impairment of cognitive function in CADASIL mice.[1] \n",
      "evidence 1: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.[2] \n",
      "evidence 2: Today, CADASIL is known to every neurologist, but the disease has not yet revealed all its secrets. A lot of effort is still needed to understand the intimate mechanisms of the disease and the most efficient targets or approaches for the development of efficient therapeutics. The history of CADASIL will be further enriched by multiple ongoing research projects worldwide, at clinical and preclinical level, and will continue to enlighten research in the field of cerebral small vessel disorders.[3] \n",
      "query: what are the potential therapeutic targets for cadasil? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Using both wild type mice and transgenic mice carrying the human mutant Notch3 gene (CADASIL mice), we have recently characterized the pathological features of CADASIL and determined the therapeutic efficacy of two hematopoietic growth factors, stem cell factor (SCF) and granulocyte colony-stimulating factor (G-CSF) in CADASIL. Our findings have revealed novel pathological changes in the endothelium of cerebral capillaries and in the neural stem cells (NSCs). We have also observed the impairment of cognitive function in CADASIL mice. [1] evidence 1: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available. [2] evidence 2: Today, CADASIL is known to every neurologist, but the disease has not yet revealed all its secrets. A lot of effort is still needed to understand the intimate mechanisms of the disease and the most efficient targets or approaches for the development of efficient therapeutics. The history of CADASIL will be further enriched by multiple ongoing research projects worldwide, at clinical and preclinical level, and will continue to enlighten research in the field of cerebral small vessel disorders. [3] query: what are the potential therapeutic targets for cadasil? summarize given evidences and query: what are the potential therapeutic targets for cadasil? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The Notch signaling pathway is a highly conserved signaling pathway that plays a key role in cell fate determination during development. The Notch signaling pathway is activated by the interaction of the Notch receptor with its ligands, Delta-like and Jagged. The Notch receptor is a single-pass transmembrane protein that is expressed on the surface of the signal-receiving cell. The Notch receptor is cleaved by a disintegrin and metalloprotease (ADAM) and γ-secretase, and the Notch intracellular domain (NICD) is released from the membrane. The NICD translocates to the nucleus and interacts with the DNA-binding protein CSL (CBF1 / Su (H) / Lag1), converting it from a transcriptional repressor to an activator. The NICD-CSL complex activates the transcription of target genes, including the Hes and Hey family of genes. [4] evidence 3: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [5] evidence 4: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [6] evidence 5: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [7] evidence 6: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [8] evidence 7: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [9] evidence 8: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 0] evidence 9: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 1] evidence 1 0: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 2] evidence 1 1: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 3] evidence 1 2: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 4] evidence 1 3: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 5] evidence 1 4: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 6] evidence 1 5: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 7] evidence 1 6: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 8] evidence 1 7: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 9] evidence 1 8: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 0] evidence 1 9: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 1] evidence 2 0: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 2] evidence 2 1: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 3] evidence 2 2: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 4] evidence 2 3: Notch']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Using both wild type mice and transgenic mice carrying the human mutant Notch3 gene (CADASIL mice), we have recently characterized the pathological features of CADASIL and determined the therapeutic efficacy of two hematopoietic growth factors, stem cell factor (SCF) and granulocyte colony-stimulating factor (G-CSF) in CADASIL. Our findings have revealed novel pathological changes in the endothelium of cerebral capillaries and in the neural stem cells (NSCs). We have also observed the impairment of cognitive function in CADASIL mice.[1] \n",
      "evidence 1: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.[2] \n",
      "evidence 2: Today, CADASIL is known to every neurologist, but the disease has not yet revealed all its secrets. A lot of effort is still needed to understand the intimate mechanisms of the disease and the most efficient targets or approaches for the development of efficient therapeutics. The history of CADASIL will be further enriched by multiple ongoing research projects worldwide, at clinical and preclinical level, and will continue to enlighten research in the field of cerebral small vessel disorders.[3] \n",
      "query: what are the potential therapeutic targets for cadasil? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Using both wild type mice and transgenic mice carrying the human mutant Notch3 gene (CADASIL mice), we have recently characterized the pathological features of CADASIL and determined the therapeutic efficacy of two hematopoietic growth factors, stem cell factor (SCF) and granulocyte colony-stimulating factor (G-CSF) in CADASIL. Our findings have revealed novel pathological changes in the endothelium of cerebral capillaries and in the neural stem cells (NSCs). We have also observed the impairment of cognitive function in CADASIL mice. [1] evidence 1: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available. [2] evidence 2: Today, CADASIL is known to every neurologist, but the disease has not yet revealed all its secrets. A lot of effort is still needed to understand the intimate mechanisms of the disease and the most efficient targets or approaches for the development of efficient therapeutics. The history of CADASIL will be further enriched by multiple ongoing research projects worldwide, at clinical and preclinical level, and will continue to enlighten research in the field of cerebral small vessel disorders. [3] query: what are the potential therapeutic targets for cadasil? summarize given evidences and query: what are the potential therapeutic targets for cadasil? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The Notch signaling pathway is a highly conserved signaling pathway that plays a key role in cell fate determination during development. The Notch signaling pathway is activated by the interaction of the Notch receptor with its ligands, Delta-like and Jagged. The Notch receptor is a single-pass transmembrane protein that is expressed on the surface of the signal-receiving cell. The Notch receptor is cleaved by a disintegrin and metalloprotease (ADAM) and γ-secretase, and the Notch intracellular domain (NICD) is released from the membrane. The NICD translocates to the nucleus and interacts with the DNA-binding protein CSL (CBF1 / Su (H) / Lag1), converting it from a transcriptional repressor to an activator. The NICD-CSL complex activates the transcription of target genes, including the Hes and Hey family of genes. [4] evidence 3: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [5] evidence 4: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [6] evidence 5: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [7] evidence 6: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [8] evidence 7: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [9] evidence 8: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 0] evidence 9: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 1] evidence 1 0: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 2] evidence 1 1: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 3] evidence 1 2: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 4] evidence 1 3: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 5] evidence 1 4: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 6] evidence 1 5: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 7] evidence 1 6: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 8] evidence 1 7: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 9] evidence 1 8: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 0] evidence 1 9: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 1] evidence 2 0: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 2] evidence 2 1: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 3] evidence 2 2: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 4] evidence 2 3: Notch']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Using both wild type mice and transgenic mice carrying the human mutant Notch3 gene (CADASIL mice), we have recently characterized the pathological features of CADASIL and determined the therapeutic efficacy of two hematopoietic growth factors, stem cell factor (SCF) and granulocyte colony-stimulating factor (G-CSF) in CADASIL. Our findings have revealed novel pathological changes in the endothelium of cerebral capillaries and in the neural stem cells (NSCs). We have also observed the impairment of cognitive function in CADASIL mice.[1] \n",
      "evidence 1: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.[2] \n",
      "evidence 2: Today, CADASIL is known to every neurologist, but the disease has not yet revealed all its secrets. A lot of effort is still needed to understand the intimate mechanisms of the disease and the most efficient targets or approaches for the development of efficient therapeutics. The history of CADASIL will be further enriched by multiple ongoing research projects worldwide, at clinical and preclinical level, and will continue to enlighten research in the field of cerebral small vessel disorders.[3] \n",
      "query: what are the potential therapeutic targets for cadasil? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Using both wild type mice and transgenic mice carrying the human mutant Notch3 gene (CADASIL mice), we have recently characterized the pathological features of CADASIL and determined the therapeutic efficacy of two hematopoietic growth factors, stem cell factor (SCF) and granulocyte colony-stimulating factor (G-CSF) in CADASIL. Our findings have revealed novel pathological changes in the endothelium of cerebral capillaries and in the neural stem cells (NSCs). We have also observed the impairment of cognitive function in CADASIL mice. [1] evidence 1: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available. [2] evidence 2: Today, CADASIL is known to every neurologist, but the disease has not yet revealed all its secrets. A lot of effort is still needed to understand the intimate mechanisms of the disease and the most efficient targets or approaches for the development of efficient therapeutics. The history of CADASIL will be further enriched by multiple ongoing research projects worldwide, at clinical and preclinical level, and will continue to enlighten research in the field of cerebral small vessel disorders. [3] query: what are the potential therapeutic targets for cadasil? summarize given evidences and query: what are the potential therapeutic targets for cadasil? < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The Notch signaling pathway is a highly conserved signaling pathway that plays a key role in cell fate determination during development. The Notch signaling pathway is activated by the interaction of the Notch receptor with its ligands, Delta-like and Jagged. The Notch receptor is a single-pass transmembrane protein that is expressed on the surface of the signal-receiving cell. The Notch receptor is cleaved by a disintegrin and metalloprotease (ADAM) and γ-secretase, and the Notch intracellular domain (NICD) is released from the membrane. The NICD translocates to the nucleus and interacts with the DNA-binding protein CSL (CBF1 / Su (H) / Lag1), converting it from a transcriptional repressor to an activator. The NICD-CSL complex activates the transcription of target genes, including the Hes and Hey family of genes. [4] evidence 3: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [5] evidence 4: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [6] evidence 5: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [7] evidence 6: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [8] evidence 7: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [9] evidence 8: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 0] evidence 9: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 1] evidence 1 0: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 2] evidence 1 1: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 3] evidence 1 2: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 4] evidence 1 3: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 5] evidence 1 4: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 6] evidence 1 5: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 7] evidence 1 6: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 8] evidence 1 7: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [1 9] evidence 1 8: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 0] evidence 1 9: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 1] evidence 2 0: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 2] evidence 2 1: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 3] evidence 2 2: Notch signaling is involved in the regulation of cell fate determination, including the differentiation of NSCs. [2 4] evidence 2 3: Notch']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Syk inhibition was found to stabilize microtubules and potentiate paclitaxel activity in cellular models of taxane-resistant ovarian cancers. We further studied the effects of Syk inhibition on paclitaxel activity in Syk(+) ovarian cancer cell models and in variants selected for taxane resistance. Syk inhibition was accomplished using RNAi and by exposure to the small molecule competitive inhibitor R406, the active metabolite of fostamatinib.[1] \n",
      "evidence 1: Therefore, inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients.[2] \n",
      "evidence 2: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.[3] \n",
      "query: would SYK inhibition be potential treatment for cadasil? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Syk inhibition was found to stabilize microtubules and potentiate paclitaxel activity in cellular models of taxane-resistant ovarian cancers. We further studied the effects of Syk inhibition on paclitaxel activity in Syk (+) ovarian cancer cell models and in variants selected for taxane resistance. Syk inhibition was accomplished using RNAi and by exposure to the small molecule competitive inhibitor R406, the active metabolite of fostamatinib. [1] evidence 1: Therefore, inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients. [2] evidence 2: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available. [3] query: would SYK inhibition be potential treatment for cadasil? summarize given evidences and query: yes. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The liver is the central organ for the metabolism of nutrients and xenobiotics. The liver is also the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug metabolism and detoxification. The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug metabolism and detoxification. The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Syk inhibition was found to stabilize microtubules and potentiate paclitaxel activity in cellular models of taxane-resistant ovarian cancers. We further studied the effects of Syk inhibition on paclitaxel activity in Syk(+) ovarian cancer cell models and in variants selected for taxane resistance. Syk inhibition was accomplished using RNAi and by exposure to the small molecule competitive inhibitor R406, the active metabolite of fostamatinib.[1] \n",
      "evidence 1: Therefore, inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients.[2] \n",
      "evidence 2: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.[3] \n",
      "query: would SYK inhibition be potential treatment for cadasil? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Syk inhibition was found to stabilize microtubules and potentiate paclitaxel activity in cellular models of taxane-resistant ovarian cancers. We further studied the effects of Syk inhibition on paclitaxel activity in Syk (+) ovarian cancer cell models and in variants selected for taxane resistance. Syk inhibition was accomplished using RNAi and by exposure to the small molecule competitive inhibitor R406, the active metabolite of fostamatinib. [1] evidence 1: Therefore, inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients. [2] evidence 2: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available. [3] query: would SYK inhibition be potential treatment for cadasil? summarize given evidences and query: yes. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The liver is the central organ for the metabolism of nutrients and xenobiotics. The liver is also the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug metabolism and detoxification. The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug metabolism and detoxification. The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n",
      "< PROMPT >\n",
      "\n",
      "evidence 0: Syk inhibition was found to stabilize microtubules and potentiate paclitaxel activity in cellular models of taxane-resistant ovarian cancers. We further studied the effects of Syk inhibition on paclitaxel activity in Syk(+) ovarian cancer cell models and in variants selected for taxane resistance. Syk inhibition was accomplished using RNAi and by exposure to the small molecule competitive inhibitor R406, the active metabolite of fostamatinib.[1] \n",
      "evidence 1: Therefore, inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients.[2] \n",
      "evidence 2: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available.[3] \n",
      "query: would SYK inhibition be potential treatment for cadasil? \n",
      "summarize given evidences and query:\n",
      "\n",
      "\n",
      "< GENERATED >\n",
      "\n",
      "['evidence 0: Syk inhibition was found to stabilize microtubules and potentiate paclitaxel activity in cellular models of taxane-resistant ovarian cancers. We further studied the effects of Syk inhibition on paclitaxel activity in Syk (+) ovarian cancer cell models and in variants selected for taxane resistance. Syk inhibition was accomplished using RNAi and by exposure to the small molecule competitive inhibitor R406, the active metabolite of fostamatinib. [1] evidence 1: Therefore, inhibition of the inflammatory macrophages would be a promising approach to attenuate NASH. In this study, we studied the implication of SYK pathway in NASH, and investigated PLGA nanoparticles-based delivery of SYK pathway inhibitor as an effective and promising therapeutic approach for the treatment of NASH. We found positive correlation between SYK expression with the pathogenesis of NASH and alcoholic hepatitis in patients. [2] evidence 2: Cerebral autosomal dominant arteriopathy with subcortical infarcts and leukoencephalopathy (CADASIL) is the most common condition of hereditary stroke and vascular dementia. CADASIL is caused by Notch3 mutation, leading to progressive degeneration of vascular smooth muscle cells (VSMCs) of the small arteries in the brain. However, the pathogenesis of CADASIL remains largely unknown, and treatment that can stop or delay the progression of CADASIL is not yet available. [3] query: would SYK inhibition be potential treatment for cadasil? summarize given evidences and query: yes. < / FREETEXT > < / PARAGRAPH > ▃ < PARAGRAPH > < FREETEXT > The liver is the central organ for the metabolism of nutrients and xenobiotics. The liver is also the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug metabolism and detoxification. The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug metabolism and detoxification. The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is also the major site of drug-induced liver injury (DILI). DILI is a major cause of acute liver failure and a major reason for drug withdrawal from the market. [4] The liver is the major site of drug metabolism and detoxification. The liver is']\n",
      "\n",
      "###############################################################################################################################################################\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(f\"generation_results_{model_name.replace('/', '-')}_zero-shot.txt\", \"w\") as results:\n",
    "    results.write(f'<< Model: {model_name} >>\\n\\n')\n",
    "    # results.write()\n",
    "\n",
    "    # for i in range(len(args)):\n",
    "    for i in range(1):\n",
    "        example, query = args #[i]\n",
    "        prompt = make_prompt(example, query, zeroshot=True)\n",
    "        \n",
    "        print('< PROMPT >\\n')\n",
    "        print(prompt)\n",
    "        results.write('< PROMPT >\\n\\n')\n",
    "        results.write(prompt)\n",
    "        \n",
    "        inputs = tokenizer(prompt, return_tensors=\"pt\", max_length=1024, truncation=True)\n",
    "        inputs.to(device)\n",
    "        outputs = model.generate(**inputs, max_length=1024)\n",
    "        generated = tokenizer.batch_decode(outputs, skip_special_tokens=True)\n",
    "        \n",
    "        print('\\n\\n< GENERATED >\\n')\n",
    "        print(generated)\n",
    "        print('\\n###############################################################################################################################################################\\n')\n",
    "        results.write('\\n\\n< GENERATED >\\n\\n')\n",
    "        results.write(*generated)\n",
    "        results.write('\\n\\n###############################################################################################################################################################\\n')\n",
    "        \n",
    "results.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# inputs = tokenizer(prompt, return_tensors=\"pt\", max_length=2048, truncation=True)\n",
    "# outputs = model.generate(**inputs, max_length=2048)\n",
    "# print(tokenizer.batch_decode(outputs, skip_special_tokens=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "mujeen_input_text = \"\"\"evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
    "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
    "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large \n",
    "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
    "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-κB. A positive regulation between NOTCH and NF-κB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
    "query: What is the therapeutic target for CADASIL? \n",
    "summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGF-b1 signaling. \n",
    "\n",
    "evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients. All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. \n",
    "evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 \n",
    "evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of vSMC . maturation of vSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of vSMC. \n",
    "query: What is the cause of the CADASIL? \n",
    "summarize given evidences and query:\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "evidence_input = \"\"\"Evidence: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms.\n",
    "Summarize the evidence: In the evidence, it is mentioned that mutations were identified in CADASIL patients and that it may result in severe disruption of the Notch3 protein.\n",
    "\n",
    "Evidence: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of vSMC . maturation of vSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of vSMC. \n",
    "Summarize the evidence:\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_text = \"\"\"evidence 0: There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with regard to downstream signaling or whether CADASIL mutations are neutral in terms of Notch signaling. Whereas the R169C mutation appeared to lead to hyperactive Notch signaling (see above), the R1031C or C455R mutations were instead shown to be hypoactive,36 \n",
    "evidence 1: We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. Support for our hypothesis is the outcome of several studies that showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic cardiovascular disease. \n",
    "evidence 2: Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused terminal regions particularly susceptible to infarcts. SMCs degeneration is followed by the emergence of large \n",
    "evidence 3: Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase activity of SYK is essential for the activation of some signaling receptor downstream effector molecules. In the present study, the activation of SYK was shown to increase the progression of peritoneal fibrosis through activation of the TGF-b1/Smad3 signaling pathway, and inhibition of TGF-b1 also resulted in down-regulation of SYK. \n",
    "evidence 4: Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage activation and identified its prominent and specific role in the activation of NF-κB. A positive regulation between NOTCH and NF-κB signaling pathway has been described previously in macrophages isolated from patients with atherosclerosis. In those patients, and in contrast with our results, \n",
    "query: What is the therapeutic target for CADASIL? \n",
    "summarize given evidences and query: Syk is a therapeutic target for CADASIL. In CADASIL, proinflammatory signaling is activated through the increase of TGFb1 signaling. The apoptosis of vascular smooth muscle cell (VSMC) and increased TGFb1 signaling cause fibrosis and vascular abnormalities in vascular epithelial cells. Syk increases fibrosis through activation of TGF-b1 signaling. \n",
    "\n",
    "evidence 0: CADASIL) is sometimes misdiagnosed as multiple sclerosis (MS). MS and CADASIL are not known to co-occur and brain magnetic resonance imagining (MRI) findings can help with differential diagnosis. Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship.\n",
    "evidence 1: Despite the availability of this information, a case report is presented of a 61-year-old woman who was misdiagnosed with MS at age 50, tested positive for CADASIL at age 56, described incorrectly as having both conditions simultaneously, and continued on MS disease-modifying medications, resulting in financial and physical hardship. Neuropsychological consultation helped initiate removal of the MS diagnosis and treatment. Better understanding is needed among clinicians that MS and CADASIL are not known to co-exist, that no association has been found between MS and the NOTCH3 mutations that cause CADASIL, and that neuroimaging and clinical features can help distinguish between the two conditions in addition to genetic testing.\n",
    "evidence 2: Antithrombotic drugs were used in three patients, in one for an unrelated coexisting prothrombotic condition. CADASIL does not seem to be associated with an unfavorable outcome of pregnancy either for women and fetuses. Patients and treating physicians should be reassured that pregnancy can be safely initiated in CADASIL, as there is no evidence to support a specific preventive antithrombotic treatment during pregnancy in CADASIL.\n",
    "query: can nilvadipine help alleviate cadasil symptoms?\n",
    "summarize given evidences and query:\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_text = \"\"\"evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. \n",
    "evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 \n",
    "evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC . maturation of vSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of vSMC.\n",
    "query: What is the cause of the CADASIL?\n",
    "summarize given evidences given the query: In evidence 0, it is mentioned that mutations were identified in CADASIL patients and that it may result in severe disruption of the Notch3 protein. In evidence 1, it is mentioned that the mutated gene was identified as NOTCH3. In evidence 2, it is mentioned that Notch3 is the first key player in the maturation of VSMCs.\n",
    "\n",
    "evidence 0: “Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy ... (Stroke 2018)” There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with …\n",
    "evidence 1: “Differences in proliferation rate between CADASIL and control vascular smooth muscle cells are ... (J Cell Mol Med 2018)” We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. … showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic …\n",
    "evidence 2: “The Role of Vascular Smooth Muscle Cells in Arterial Remodeling: Focus on Calcification ... (Int J Mol Sci 2019)” Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused …\n",
    "query: What is the therapeutic target for CADASIL?\n",
    "summarize given evidences given the query: \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_text = \"\"\"Evidence 0: Here we report the characterization of the human Notch3 gene which we mapped to the CADASIL critical region. We have identified mutations in CADASIL patients that cause serious disruption of this gene, indicating that Notch3 could be the defective protein in CADASIL patients.All these missense mutations may result in severe disruption of the Notch3 protein, as suggested by the highly conserved nature of the aminoacid residues involved, particularly the cysteines that are key features of EGF likedomains24. These results indicate that these nucleotide substitutions are pathogenic mutations rather than rare polymorphisms. \n",
    "Evidence 1: Linkage studies in other families enabled further refinement of this genetic interval16,17 and identification of the mutated gene as NOTCH3 (Notch homolog 3).18 \n",
    "Evidence 2: CADASIL, a hereditary vascular dementia, suggesting a role for Notch3 in vessel homeostasis (Joutel et al. 1996). CADASIL is a late-onset disorder, and neurological symptoms arise from a slowly developing systemic vasculopathy, characterized ultimately by degeneration of VSMC . maturation of vSMC, and ends around P28 when the artery acquires its final shape. We identify Notch3 to be the first key player of this process, by regulating cell-autonomously the arterial differentiation and maturation of vSMC.\n",
    "Query: What is the cause of the CADASIL?\n",
    "Summarization of evidences given the query:\n",
    "CADASIL is a rare hereditary cerebrovascular disease caused by a NOTCH3 mutation[1][2]. NOTCH3 is particularly important for the maturation of vascular smooth muscle cells (VSMCs). A decreased VSMC is a pathological hallmark for CADASIL[3].\n",
    "\n",
    "Evidence 0: “Role of NOTCH3 Mutations in the Cerebral Small Vessel Disease Cerebral Autosomal Dominant Arteriopathy … (Stroke 2018)” There is currently no consensus on whether CADASIL mutations generate hyperactive or hypoactive Notch3 proteins with …\n",
    "Evidence 1: “Differences in proliferation rate between CADASIL and control vascular smooth muscle cells are … (J Cell Mol Med 2018)” We believe that increased TGFb3 reflects an inflammatory condition and even an involvement of TGFb in fibrosis in CADASIL. … showed NOTCH3 and TGFb1 signalling play a key role in the pathogenesis and progression of chronic …\n",
    "Evidence 2: “The Role of Vascular Smooth Muscle Cells in Arterial Remodeling: Focus on Calcification … (Int J Mol Sci 2019)” Over time, VSMCs apoptosis leads to fibrosis and thickening of the arterial wall, progressive lumen stenosis and vascular insufficiency that makes the already poorly perfused …\n",
    "Evidence 3: “Spleen Tyrosine Kinase (SYK) in the Progression of Peritoneal Fibrosis Through Activation of the … (Med Sci Monit 2019)” Previously published studies have shown that TGF-b signaling is closely associated with the activity of SYK, and the kinase …\n",
    "Evidence 4: “NOTCH3 signaling is essential for NF-κB activation in TLRactivated macrophages (Sci Rep 2020)” Our work confirms and highlights the relevance of NOTCH3 expression and signaling in pro-inflammatory macrophage …\n",
    "Evidence 5: “Niclosamide repurposed for the treatment of inflammatory airway disease (JCI Insight 2019)” Niclosamide and derivatives — potent inhibitors of TMEM16A/F and Ca2+ signaling. NFA is a nonsteroidal antiinflammatory compound that inhibits goblet …\n",
    "Evidence 6: “The Ca2+-gated channel TMEM16A amplifies capillary pericyte contraction and reduces … (J Clin Invest. 2022)” The key findings of this study are that (a) Cl– fluxes mediated by the Ca2+-gated channel TMEM16A are a crucial determinant of pericyte tone\n",
    "Query: What is the therapeutic target for CADASIL?\n",
    "Summarization of evidences given the query: \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biogpt",
   "language": "python",
   "name": "biogpt"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
